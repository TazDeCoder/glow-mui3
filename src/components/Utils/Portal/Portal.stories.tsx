import type { ComponentMeta, ComponentStory } from '@storybook/react';

import { Portal } from './Portal';
import { Box } from '../Box';

export default {
  title: 'Components/Utils/Portal',
  component: Portal,
} as ComponentMeta<typeof Portal>;

export const Default: ComponentStory<typeof Portal> = () => (
  <>
    <Portal>
      <Box p={1} my={1} bg="primary.main" color="primary.onMain">
        Portal 1
      </Box>
    </Portal>
    <Portal>
      <Box p={1} my={1} bg="secondary.main" color="secondary.onMain">
        Portal 2
      </Box>
    </Portal>
  </>
);
