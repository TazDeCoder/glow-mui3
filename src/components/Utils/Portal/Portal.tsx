import { useState, useLayoutEffect } from 'react';
import { createPortal } from 'react-dom';

export interface PortalProps {
  /** Content of the container element */
  children: React.ReactNode;
  /** Container element to render as parent
   * @default div
   */
  node?: HTMLElement;
}

export const Portal: React.FC<PortalProps> = ({ children, node }) => {
  const [defaultNode] = useState(() => node || document.createElement('div'));

  useLayoutEffect(() => {
    if (defaultNode) {
      document.body.appendChild(defaultNode);
    }

    return () => {
      if (defaultNode) {
        document.body.removeChild(defaultNode);
      }
    };
  }, [node, defaultNode]);

  // @ts-ignore
  return createPortal(children, defaultNode);
};
