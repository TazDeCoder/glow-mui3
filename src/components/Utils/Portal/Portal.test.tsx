import { render } from '@testing-library/react';

import { Portal } from './Portal';

describe('<Portal />', () => {
  describe('Snapshots', () => {
    it('renders correctly', () => {
      const { container } = render(<Portal>Foo</Portal>);
      expect(container).toMatchSnapshot();
    });
  });
});
