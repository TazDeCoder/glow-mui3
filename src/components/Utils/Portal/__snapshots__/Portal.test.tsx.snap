// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Portal /> Snapshots renders correctly 1`] = `<div />`;
